stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -3318,44 +3347,106 @@ static void _gatherWrittenGenericArgs(
     auto lhsFlatIndex =
       _depthIndexToFlatIndex(lhsParam->first, lhsParam->second,
                              genericParamCounts);
-    if (!lhsFlatIndex || *lhsFlatIndex >= allGenericArgs.size())
-      continue;
+    if (!lhsFlatIndex || *lhsFlatIndex >= genericArgs.size())
+      return false;
 
-    if (!allGenericArgs[*lhsFlatIndex]) {
+    if (!genericArgs[*lhsFlatIndex]) {
       // Substitute into the right-hand side.
-      SubstGenericParametersFromWrittenArgs substitutions(allGenericArgs,
-                                                          genericParamCounts);
-      allGenericArgs[*lhsFlatIndex] =
-          MetadataOrPack(swift_getTypeByMangledName(MetadataState::Abstract,
+      auto *genericArg =
+          swift_getTypeByMangledName(MetadataState::Abstract,
             req.getMangledTypeName(),
-            (const void * const *)allGenericArgs.data(),
+            keyArgs.data(),
             [&substitutions](unsigned depth, unsigned index) {
               return substitutions.getMetadata(depth, index).Ptr;
             },
             [&substitutions](const Metadata *type, unsigned index) {
               return substitutions.getWitnessTable(type, index);
-            }).getType().getMetadata());
+            }).getType().getMetadata();
+      if (!genericArg)
+        return false;
+
+      genericArgs[*lhsFlatIndex] = MetadataOrPack(genericArg);
       continue;
     }
 
     // If we do have an argument for this parameter, it might be that
     // the right-hand side is itself a generic parameter, which means
     // we have a same-type constraint A == B where A is already filled in.
     auto rhsParam = demangleToGenericParamRef(req.getMangledTypeName());
-    if (!rhsParam)
+
+    // If the rhs parameter is not a generic parameter itself with
+    // (depth, index), it could potentially be some associated type. Check it
+    // again later once we've found all of the other same types.
+    if (!rhsParam) {
+      needToCheckAssociatedSameTypes = true;
       continue;
+    }
 
     auto rhsFlatIndex =
       _depthIndexToFlatIndex(rhsParam->first, rhsParam->second,
                              genericParamCounts);
-    if (!rhsFlatIndex || *rhsFlatIndex >= allGenericArgs.size())
+    if (!rhsFlatIndex || *rhsFlatIndex >= genericArgs.size())
+      return false;
+
+    if (genericArgs[*rhsFlatIndex] || !genericArgs[*lhsFlatIndex])
+      return false;
+
+    genericArgs[*rhsFlatIndex] = genericArgs[*lhsFlatIndex];
+  }
+
+  if (!needToCheckAssociatedSameTypes) {
+    return true;
+  }
+
+  // Once again, loop through our list and look for same type constraints where
+  // the rhs is an associated type of sorts.
```

Comment:
I believe it should be possible to populate all genericArgs in one pass. A generic parameter is non-canonical if you have a requirement `T == ConcreteType` or `U == T` (note the order) where `U` is another generic parameter.

`T.A.B == U` does not make `U` non-canonical on the other hand, so it should not be necessary to populate `U` here.